Move IHealth to mod code

reaperrr opened the pull request:
Adds `IActorDeath` interface housing `IsDead` for backwards-compatibility, moves `IHealth(Info)` interfaces as well as `GetDamageState()` and `Kill` extensions to Mods.Common, gets rid of `Actor.InflictDamage` entirely.

I thought about splitting into multiple PRs, but none of the commits are really huge or complex. Reviewing commit by commit highly recommended.
Commit order may look a bit confusing though (here on gh at least), since I reordered and updated some commits right before pushing.

reaperrr committed:
Add IActorDeath interface

reaperrr committed:
Move GetDamageState() from Game to Mod code

in preparation of removing IHealth from OpenRA.Game.

reaperrr committed:
Use already cached Health where possible

reaperrr committed:
Cache Health where ActorExts may be too expensive

Cache Health look-ups in places where using
ActorExts would potentially be too expensive
once we move them to mod code.

reaperrr committed:
Move Kill from Actor to mod ActorExts

Shouldn't be as performance-sensitive
as GetDamageState, so updating cases
that call it to cache Health shouldn't
be necessary.

reaperrr committed:
Remove Actor.InflictDamage

No ActorExts replacement for this, because
a) potentially VERY performance-sensitive
b) not that many places were affected
(of which some were already caching Health anyway).

reaperrr committed:
Cache Health in HitShape

To avoid separate Health look-ups in warheads
once Actor.InflictDamage is removed.

reaperrr committed:
Remove IHealth caching from Actor

reaperrr committed:
Pass HP/DamageState to FAL directly

Instead of looking up IHealth in FAL.

reaperrr committed:
Move IHealth interfaces to Mods.Common

reaperrr commented on OpenRA.Game/Traits/Player/FrozenActorLayer.cs:
```diff
@@ -102,29 +101,25 @@ public FrozenActor(Actor actor, ICreatesFrozenActors frozenTrait, PPos[] footpri
 		public bool IsValid { get { return Owner != null; } }
 		public ActorInfo Info { get { return actor.Info; } }
-		public Actor Actor { get { return !actor.IsDead ? actor : null; } }
+		public Actor Actor { get { return !actor.Disposed ? actor : null; } }
```
Forgot to revert this, will fix it once I next update this branch.

reaperrr commented:
Updated 1a9c2c1 with two changes:
- Cache `IHealth` instead of `Health` where possible, I guess we'll probably want to keep the interface for downstream mods
- ~Instead of each `WithSpriteBody`, `WithSpriteTurret` and `WithSpriteBarrel` caching `IHealth` individually, it's now cached in `RenderSprites` and publicly returning `DamageState` from there.~

pchote commented:
I'm not sure about `RenderSprites` having a concept of health - it seems completely reasonable for a mod to define an actor without it.

reaperrr commented:
> I'm not sure about RenderSprites having a concept of health - it seems completely reasonable for a mod to define an actor without it.

Reverted that part.

reaperrr commented:
I'm a bit stumped here. Starting from 96c0bb7 this triggers #17926, but I just cannot figure out why. It doesn't appear to be trait construction order, since priority-loading it `OccupiesSpace`-style didn't help. Explicitly implementing `IActorDeath.IsDead` in `Health` didn't help, either.

reaperrr commented:
Oh wow... the only reason the "win" trigger works in allies01 is because the heli is being disposed directly instead of killed, but `Actor.Disposed` isn't `true` yet because of being set in a FrameEndTask. What broke it was that I made `IsDead` return `true` on `WillDispose` as well.

A trigger relying on that timing gap sounds highly fishy to me. Should I really revert returning `IsDead` on `WillDispose`, or should we rather fix the triggers that rely on the old behavior?

pchote commented:
lets fix that and any other missions first.

reaperrr committed:
Streamline SelectionBars IsDead checks

They already check Actor.IsDead a few lines above,
so no need to check health.IsDead as well.

reaperrr commented:
Removed `WillDispose` check anyway, as discussed on IRC.
Also added minor perf optimization of `SelectionBarAnnotationRenderable`s.

abcdefg30 commented:
Yes, a lot of missions discard actors since they need to be removed from the world without being killed (since that would be bogus). Breaking all `OnRemovedFromWorld` triggers sounds like a bad idea imho.

pchote commented:
What is the current status here wrt mission regressions?

reaperrr commented:
> What is the current status here wrt mission regressions?

_Shouldn't_ (fingers crossed) be an issue anymore due to

> Removed `WillDispose` check anyway, as discussed on IRC.

pchote commented on OpenRA.Mods.Common/ActorExts.cs:
```diff
@@ -82,5 +82,24 @@ public static CPos ClosestCell(this Actor self, IEnumerable<CPos> cells)
+			if (self.WillDispose)
+				return DamageState.Dead;
+
+			var health = self.TraitOrDefault<Health>();
```
We could avoid these queries by making `Actor.death` public and doing `self.death as IHealth` - but this looks awkward so we should try and come up with a different name for the interface / field. Maybe something to do with invalidation?

abcdefg30 commented:
@reaperrr what is the status here?

reaperrr commented:
I currently don't have the energy to deal with the (seemingly major) rebase and thinking about the interface name, I'll reopen if/when I get back to this.
